### PR TITLE
feat(ecs): add support for ProxyConfiguration in ECS TaskDefinition

### LINKED ISSUE #3977: ECS: Property 'ProxyConfiguration' is missing in type 'TaskDefinition'
The ProxyConfiguration property supported by CloudFormation is not available in the TaskDefinition construct.

### PR DESCRIPTION
**Pull request by iamhopaul123**
Add proxyConfiguration support for ECS task definition. 

Addresses #3977 

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+      if (!props.properties.ignoredUID && !props.properties.ignoredGID) {
+        throw new Error("Either ignoredUID or ignoredGID should be specified.");
+      }
+    }
+  }
+
+  /**
+   * Called when the proxy configuration is configured on a task definition.
+   */
+  public bind(): CfnTaskDefinition.ProxyConfigurationProperty {
```
[the remark this answers is not in the document]
Comment (iamhopaul123) [reply packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts]: Agree

File: packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+ * AMI version 20190301 or later, then they contain the required versions of the container agent and ecs-init.
+ * For more information, see [Amazon ECS-optimized AMIs](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-optimized_AMI.html).
+ *
+ * For tasks using the Fargate launch type, the task or service requires platform version 1.3.0 or later.
+ */
+export class AppMeshProxyConfiguration extends ProxyConfiguration {
+  /**
+   * Constructs a new instance of the AppMeshProxyConfiguration class.
+   */
+  constructor(private readonly props: AppMeshProxyConfigurationConfigProps) {
```
[the remark this answers is not in the document]
Comment (iamhopaul123) [reply packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts]: Yeah it works without properties being set.

File: packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+          },
+          {
+            name: "EgressIgnoredIPs",
+            value: String(configProps.egressIgnoredIPs)
+          }
+        ]),
+        type: configType
+      };
+    }
+    return {
```
[the remark this answers is not in the document]
Comment (iamhopaul123) [reply packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts]: :+1:

File: packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+  /**
+   * Called when the proxy configuration is configured on a task definition.
+   */
+  public bind(): CfnTaskDefinition.ProxyConfigurationProperty {
+    const configProps = this.props.properties;
+    const configType = "APPMESH";
+    if (configProps) {
+      return {
+        containerName: this.props.containerName,
+        proxyConfigurationProperties: removePropsEmpty([
```
[the remark this answers is not in the document]
Comment (iamhopaul123) [reply packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts]: I think according to the doc you referred to and this [doc](https://docs.aws.amazon.com/app-mesh/latest/userguide/mesh-getting-started-ecs.html), at least EgressIgnoredPorts/IP can be empty and users do not have to define that. According to my experiment, I tried set EgressIgnoredPorts and IP as `undefined` and it worked well.

File: packages/@aws-cdk/aws-ecs/lib/proxy-configuration/generic-proxy-configuration.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
+import { CfnTaskDefinition } from '../ecs.generated';
+import { ProxyConfiguration } from './proxy-configuration';
+
+/**
+ * The generic configuration to use when setting a proxy configuration.
+ */
+export interface GenericProxyConfigurationConfigProps {
```
[the remark this answers is not in the document]
Comment (iamhopaul123) [reply packages/@aws-cdk/aws-ecs/lib/proxy-configuration/generic-proxy-configuration.ts]: Removed.

File: packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (pkandasamy91) [reply packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts]: Are you able to create a stack and use AppMesh?

File: packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+  /**
+   * Specifies the port that outgoing traffic from the AppPorts is directed to.
+   */
+  readonly proxyEgressPort: number;
+
+  /**
+   * The egress traffic going to these specified ports is ignored and not redirected to the ProxyEgressPort. It can be an empty list.
+   */
+  readonly egressIgnoredPorts?: number[];
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts]: From what I can see, this is required but can be an empty list.

File: packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+  /**
+   * The egress traffic going to these specified IP addresses is ignored and not redirected to the ProxyEgressPort. It can be an empty list.
+   */
+  readonly egressIgnoredIPs?: string[];
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts]: Same comment as above.

File: packages/@aws-cdk/aws-ecs/test/test.app-mesh-proxy-configuration.ts [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+          appPorts: [80, 81],
+          proxyIngressPort: 80,
+          proxyEgressPort: 81
+        }
+      })});
+    }, /Either ignoredUID or ignoredGID should be specified./);
+
+    test.done();
+  }
+};
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs/test/test.app-mesh-proxy-configuration.ts]: Can you add a test for if ignoredPorts/Ips is empty?

File: packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+ */
+export class AppMeshProxyConfiguration extends ProxyConfiguration {
+  /**
+   * Constructs a new instance of the AppMeshProxyConfiguration class.
+   */
+  constructor(private readonly props: AppMeshProxyConfigurationConfigProps) {
+    super();
+    if (props.properties) {
+      if (!props.properties.ignoredUID && !props.properties.ignoredGID) {
+        throw new Error("Either ignoredUID or ignoredGID should be specified.");
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts]: nit: You might want to reword this as "At least one of ignoredUID or ignoredGID should be specified"

Comment (iamhopaul123) [reply packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts]: Done!

**Comment by iamhopaul123** [comment]
This is an app mesh [example](https://github.com/aws/aws-app-mesh-examples/blob/master/examples/apps/colorapp/ecs/colorteller-base-task-def.json#L3-L28), which can be generated by this PR.

File: packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (pkandasamy91) [reply packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts]: You cannot call removePropsEmpty if those properties need to be populated.

[the remark this answers is not in the document]
Comment (iamhopaul123) [reply packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts]: Actually "AppPorts", "ProxyIngressPort", "ProxyEgressPort" these three properties are required for users, so that they cannot be empty and won't be removed by `removePropsEmpty`. As for "EgressIgnoredPorts", "EgressIgnoredIPs" after talking to agent teams, they looked at the agent code and verified that those two fields are not required and they don't do any validation for their existence.

File: packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+
+  /**
+   * Called when the proxy configuration is configured on a task definition.
+   */
+  public bind(_scope: Construct, _taskDefinition: TaskDefinition): CfnTaskDefinition.ProxyConfigurationProperty {
+    const configProps = this.props.properties;
+    const configType = "APPMESH";
+    return {
+      containerName: this.props.containerName,
+      proxyConfigurationProperties: removePropsEmpty([
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts]: This logic should move into a render method (similar to how containerDefinition has renderContainerDefinition) which will do lazy values and omit empty arrays.

File: packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+          name: "ProxyIngressPort",
+          value: String(configProps.proxyIngressPort)
+        },
+        {
+          name: "ProxyEgressPort",
+          value: String(configProps.proxyEgressPort)
+        },
+        {
+          name: "EgressIgnoredPorts",
+          value: String((configProps.egressIgnoredPorts && configProps.egressIgnoredPorts.length) ? configProps.egressIgnoredPorts : undefined)
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts]: This logic doesn't seem the best way to check for an empty array. You might want to use Lazy to omitEmptyArrays will solve your issue here.

File: packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+        },
+        {
+          name: "EgressIgnoredIPs",
+          value: String((configProps.egressIgnoredIPs && configProps.egressIgnoredIPs.length) ? configProps.egressIgnoredIPs : undefined)
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts]: Same as above.

File: packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts [hunk 13]
```diff
@@ -92,50 +92,20 @@ export class AppMeshProxyConfiguration extends ProxyConfiguration {
...
-  const newPairs = [];
-  for (const pair of pairs) {
-    if (pair.value !== "undefined") {
-      newPairs.push(pair);
+function renderProperties(props: AppMeshProxyConfigurationProps): CfnTaskDefinition.KeyValuePairProperty[] {
+  const ret = [];
+  for (const [k, v] of Object.entries(props)) {
+    const key = String(k);
+    const value = String(v);
+    if (value !== "undefined" && value !== "") {
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts]: Wouldn't this mean that the EgressIgnoredPorts and EgressIgnoredIPs will not be written if they're undefined? Is this the behaviour we want?

Comment (iamhopaul123) [reply packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts]: Yes it is. This is the current behavior https://docs.aws.amazon.com/app-mesh/latest/userguide/mesh-getting-started-ecs.html and the agent team will continue support this and also support put an empty string as the `value` to do the same thing in the future.

File: packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+  }
+}
+
+function renderProperties(props: AppMeshProxyConfigurationProps): CfnTaskDefinition.KeyValuePairProperty[] {
+  const ret = [];
+  for (const [k, v] of Object.entries(props)) {
+    const key = String(k);
+    const value = String(v);
+    if (value !== "undefined" && value !== "") {
+      ret.push({ ["name"]: key.charAt(0).toUpperCase() + key.slice(1), ["value"]: value });
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts]: You should use: https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/core/lib/util.ts#L10 to capitalize instead of your own logic, to maintain consistency

Comment (pkandasamy91) [reply packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts]: I'd rather you moved out the logic to construct the name into it's own const.

File: packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+    };
+  }
+}
+
+function renderProperties(props: AppMeshProxyConfigurationProps): CfnTaskDefinition.KeyValuePairProperty[] {
+  const ret = [];
+  for (const [k, v] of Object.entries(props)) {
+    const key = String(k);
+    const value = String(v);
+    const capitalizedKey = key.charAt(0).toUpperCase() + key.slice(1);
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs/lib/proxy-configuration/app-mesh-proxy-configuration.ts]: You want to do this in the If check, otherwise you're doing additional work for no reason.
